From 43198bbd7f2fb08bb6e85c89f8220f887ed13261 Mon Sep 17 00:00:00 2001
From: Gary Bisson <gary.bisson@ezurio.com>
Date: Wed, 12 Mar 2025 11:25:49 +0100
Subject: [PATCH 1/2] media: imx: isi-cap: add pivariety workaround

Per arducam, their sensor must be turned on after the CSI controller.

Signed-off-by: Gary Bisson <gary.bisson@ezurio.com>

Upstream-Status: Inappropriate NXP proprietary driver
---
 drivers/staging/media/imx/imx8-isi-cap.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/staging/media/imx/imx8-isi-cap.c b/drivers/staging/media/imx/imx8-isi-cap.c
index c3d6bbeed6a2..0ca095866b2c 100644
--- a/drivers/staging/media/imx/imx8-isi-cap.c
+++ b/drivers/staging/media/imx/imx8-isi-cap.c
@@ -120,6 +120,7 @@ static int mxc_isi_pipeline_enable(struct mxc_isi_cap_dev *isi_cap, bool enable)
 	struct media_device *mdev = entity->graph_obj.mdev;
 	struct media_graph graph;
 	struct v4l2_subdev *subdev;
+	struct v4l2_subdev *late_subdev = NULL;
 	int ret = 0;
 
 	mutex_lock(&mdev->graph_mutex);
@@ -148,12 +149,24 @@ static int mxc_isi_pipeline_enable(struct mxc_isi_cap_dev *isi_cap, bool enable)
 			continue;
 		}
 
+		if (strstr(entity->name, "pivariety")) {
+			late_subdev = subdev;
+			continue;
+		}
+
 		ret = v4l2_subdev_call(subdev, video, s_stream, enable);
 		if (ret < 0 && ret != -ENOIOCTLCMD) {
 			dev_err(dev, "subdev %s s_stream failed\n", subdev->name);
 			break;
 		}
 	}
+
+	if (late_subdev) {
+		ret = v4l2_subdev_call(late_subdev, video, s_stream, enable);
+		if (ret < 0 && ret != -ENOIOCTLCMD)
+			dev_err(dev, "subdev %s s_stream failed\n", subdev->name);
+	}
+
 	mutex_unlock(&mdev->graph_mutex);
 	media_graph_walk_cleanup(&graph);
 
-- 
2.53.0

